src/librustc/mir/interpret/mod.rs

```diff
@@ -254,17 +256,281 @@ pub struct Allocation {
+                trace!("write_value_to_ptr valpair: {:#?}", layout);
+                let (a, b) = match layout.abi {
+                    layout::Abi::ScalarPair(ref a, ref b) => (&a.value, &b.value),
+                    _ => bug!("from_value: invalid ByValPair layout: {:#?}", layout)
+                };
+                let (a_size, b_size) = (a.size(tcx), b.size(tcx));
+                let b_offset = a_size.abi_align(b.align(tcx)).bytes();
+
+                let a_padding = b_offset - a_size.bytes();
+                let b_padding = layout.size.bytes() - (b_offset + b_size.bytes());
+
+                let mut a_mem = Allocation::from_primval(
+                    tcx,
+                    a_val,
+                    a_size.bytes(),
+                    a.align(tcx));
+                let b_mem = Allocation::from_primval(
+                    tcx,
+                    b_val,
+                    b_size.bytes(),
+                    b.align(tcx));
+
+                a_mem.extend(&Allocation::undef(a_padding, byte_align));
+                a_mem.extend(&b_mem);
+                a_mem.extend(&Allocation::undef(b_padding, byte_align));
+
+                a_mem
+            }
+        }
+    }
+
+    pub fn from_primval(
+        tcx: TyCtxt<'_, '_, '_>,
+        val: PrimVal,
+        size: u64,
+        align: Align,
+    ) -> Self {
+        match val {
+            PrimVal::Ptr(val) => {
+                assert_eq!(size, tcx.data_layout.pointer_size.bytes());
+                let mut alloc = Allocation::from_primval(
+                    tcx,
+                    PrimVal::Bytes(val.offset as u128),
+                    size,
+                    align);
+                alloc.relocations.insert(0, val.alloc_id);
+                alloc
+            }
+
+            PrimVal::Bytes(bytes) => {
+                assert!(size <= 16);
+                let dst = &mut [0; 16];
+                let dst = &mut dst[0..(size as usize)];
+                let endianness = tcx.data_layout.endian;
+                write_target_uint(endianness, dst, bytes).unwrap();
+                Allocation::from_bytes(dst, align)
+            }
+
+            PrimVal::Undef => Allocation::undef(size, align),
+        }
+    }
+
+    pub fn from_primval_ty<'tcx>(
+        tcx: TyCtxt<'_, 'tcx, '_>,
+        val: PrimVal,
+        param_env: ty::ParamEnv<'tcx>,
+        ty: Ty<'tcx>,
+    ) -> Self {
+        let layout = tcx.global_tcx().layout_of(param_env.and(ty)).unwrap();
+        Allocation::from_primval(tcx, val, layout.size.bytes(), layout.align)
+    }
+
+    // FIXME: Allow conversions to a Value pair
+    pub fn to_primval(
+        &self,
+        tcx: TyCtxt<'_, '_, '_>,
+        size: u64,
+        align: Align,
+    ) -> Option<PrimVal> {
+        if self.bytes.len() as u64 != size {
+            return None;
+        }
+        // This mirrors the logic of rustc_mir::interpret::memory::Memory::read_primval
```

Comment:
can `Memory::read_primval` just forward to this function?

Reply:
That question also applies to the other functions here. Can we move functionality from memory here by forwarding from the memory calls?

Reply:
> can Memory::read_primval just forward to this function?

This doesn't not read from an offset, so no. The other way around could work though.

I want to split Allocation into two types though, one for constant which is interned and read-only and one for Miri to use, so I don't think we should do this refactoring.

Reply:
So the deduplication of code will happen with the split into two types?

Reply:
No, it just won't be possible to deduplicate it.

Reply:
Can't you create a function that takes an offset, and have this function forward to it by filling out the offset with `0`?

Reply:
We might be able to use Miri methods here if we move the rest of the compiler away from using `Value` and `PrimVal`.

Reply:
Hmmm I don't understand this argument. Can't miri call this method or the hypothetical one with offsets?

Reply:
This code is now gone.